### PR TITLE
Change image option form builder to defined form builder

### PR DESCRIPTION
**Pull request by djbuch**
Change image option form builder to a defined form builder comming from PrestaShop\PrestaShop\Core\Form\IdentifiableObject\Builder\FormBuilder in order to be able to use the hook to add custom fields.

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Change image option form builder to a defined form builder coming from PrestaShop\PrestaShop\Core\Form\IdentifiableObject\Builder\FormBuilder in order to be able to use the hook to add custom fields. And add the actionBeforeUpdateFormImageFormHandler and actionAfterUpdateFormImageFormHandler hooks in the AdminProductWrapper
| Type?         | improvement
| Category?     | BO
| BC breaks?    | yes : the hookDispatcher was added to the AdminProductWrapper
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | Create a module using the actionFormImageFormBuilderModifier and actionAfterUpdateFormImageFormHandler hook to allow having new fields in the image form<br>A module has been sent to @SimonGrn 

:notebook: **BC Break**
* A 4th parameter `HookDispatcherInterface $hookDispatcher` has been added to the class `PrestaShop\PrestaShop\Adapter\Product\AdminProductWrapper` in `src/Adapter/Product/AdminProductWrapper.php`

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/20864)

### CODE REVIEW

File: src/Core/Form/IdentifiableObject/DataProvider/ImageFormDataProvider.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
```
Comment (sowbiba) [inline src/Core/Form/IdentifiableObject/DataProvider/ImageFormDataProvider.php:26]:
```suggestion
```suggestion
declare(strict_types=1);
```
```

Comment (sowbiba) [inline src/Core/Form/IdentifiableObject/DataProvider/ImageFormDataProvider.php:27]:
```suggestion
declare(strict_types=1);

```

File: src/PrestaShopBundle/Form/Admin/Catalog/Image/FormImageType.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
```
Comment (sowbiba) [inline src/PrestaShopBundle/Form/Admin/Catalog/Image/FormImageType.php:27]:
```suggestion
declare(strict_types=1);

```

File: src/Adapter/Product/AdminProductWrapper.php [hunk 3]
```diff
@@ -83,20 +84,27 @@ class AdminProductWrapper
...
      * Constructor : Inject Symfony\Component\Translation Translator.
      *
      * @param object $translator
      * @param array $employeeAssociatedShops
      * @param Locale $locale
      * @param FloatParser|null $floatParser
+     * @param HookDispatcher $hookDispatcher
      */
-    public function __construct($translator, array $employeeAssociatedShops, Locale $locale, FloatParser $floatParser = null)
+    public function __construct($translator, array $employeeAssociatedShops, Locale $locale, FloatParser $floatParser = null, HookDispatcherInterface $hookDispatcher)
```
Comment (PierreRambaud) [inline src/Adapter/Product/AdminProductWrapper.php:100]: It's a BC break, can you explain it in your PR description please?

File: src/Adapter/Product/AdminProductWrapper.php [hunk 4]
```diff
@@ -870,13 +878,21 @@ public function ajaxProcessUpdateImagePosition($data)
...
         $img = new Image((int) $idImage);
         if ($data['cover']) {
             Image::deleteCover((int) $img->id_product);
             $img->cover = 1;
         }
         $img->legend = $data['legend'];
         $img->update();
+        $this->hookDispatcher->dispatchWithParameters('actionAfterUpdateFormImageFormHandler', [
+            'id' => (int) $idImage,
+            'form_data' => &$data,
```
Comment (PierreRambaud) [inline src/Adapter/Product/AdminProductWrapper.php:893]:
```suggestion
            'form_data' => &$data,
            'img' => $img,
```

File: src/PrestaShopBundle/Form/Admin/Catalog/Image/FormImageType.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+class FormImageType extends TranslatorAwareType
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('legend', 'PrestaShopBundle\Form\Admin\Type\TranslateType', [
+                'type' => 'Symfony\Component\Form\Extension\Core\Type\TextareaType',
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Form/Admin/Catalog/Image/FormImageType.php]:
```suggestion
                'type' => TextareaType::class,
```
and use namespace :)

File: src/PrestaShopBundle/Form/Admin/Catalog/Image/FormImageType.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ */
+class FormImageType extends TranslatorAwareType
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('legend', 'PrestaShopBundle\Form\Admin\Type\TranslateType', [
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Form/Admin/Catalog/Image/FormImageType.php]:
```suggestion
            ->add('legend', TranslateType::class, [
```

File: src/Adapter/Product/AdminProductWrapper.php (hunk 3 above)
Comment (djbuch) [reply src/Adapter/Product/AdminProductWrapper.php:100]: Oh yes, didn't notice this.

Should I leave this BC or should I implement the hook call the "old" way ?

File: src/PrestaShopBundle/Form/Admin/Catalog/Image/FormImageType.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+                'type' => 'Symfony\Component\Form\Extension\Core\Type\TextareaType',
+                'options' => [],
+                'locales' => $this->locales,
+                'hideTabs' => true,
+                'label' => $this->trans('Caption', 'Admin.Catalog.Feature'),
+                'required' => false,
+            ])
+            ->add('cover', 'Symfony\Component\Form\Extension\Core\Type\CheckboxType', [
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Form/Admin/Catalog/Image/FormImageType.php]:
```suggestion
            ->add('cover', CheckboxType::class, [
```

File: src/Adapter/Product/AdminProductWrapper.php (hunk 3 above)
Comment (PierreRambaud) [reply src/Adapter/Product/AdminProductWrapper.php:100]: I prefer the BC than using singleton :)

Comment (djbuch) [reply src/Adapter/Product/AdminProductWrapper.php:100]: OK, then I will also change all the other hook call in this file so it will be cleaner

**Comment by Progi1984** [comment]
@djbuch Could you provide a module or provide a scenario with ps_qualityassurance like in #20207 for testing hooks ? Thanks

**Review by Progi1984 — APPROVED** [review]

**Comment by djbuch** [comment]
Hello, how could I send the module ? The module I could provided should not be available in open-source, it was developed for a customer of our web agency.

**Comment by Progi1984** [comment]
@djbuch Do you accept to send it to @PrestaShop/qa-functional for internal testing ?

If yes, send it with scenario at : `quality@prestashop.com`. Please indicate this PR.

**Comment by djbuch** [comment]
Well not really, got this mail in return


We're writing to let you know that the group you tried to contact (quality) may not exist, or you may not have permission to post messages to the group. A few more details on why you weren't able to post:

* You might have spelled or formatted the group name incorrectly.
* The owner of the group may have removed this group.
* You may need to join the group before receiving permission to post.
* This group may not be open to posting.

If you have questions related to this or any other Google Group, visit the Help Center at https://support.google.com/a/prestashop.com/bin/topic.py?topic=25838.

Thanks,

prestashop.com admins

**Comment by Progi1984** [comment]
Sorry @djbuch. Send it to the lead QA Manager : `simon.garny@prestashop.com`. Thanks

**Comment by djbuch** [comment]
OK done

**Review by sowbiba — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by SD1982** [comment]
LGTM the new field seems to be correctly added 
image
But i have an empty space between the images and the bottom of the block thant i don't have without the PR( the module is installed too)

**With the PR**
image

**Without the PR**
image

Maybe related with the conflicting file ?

**Comment by sarahdib** [comment]
Hello

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to ask to re-open when it will be possible to fix the conflicting file

Thank you!

**Pull request closed without merging**